tree/tree/src/TTreeCacheUnzip.cxx

```diff
@@ -679,241 +480,379 @@ void TTreeCacheUnzip::ResetCache()
-            }
+   char *ptr = 0;
+   Int_t loclen = UnzipBuffer(&ptr, locbuff);
+   if ((loclen > 0) && (loclen == objlen+keylen)) {
+           if ( (myCycle != fCycle)  || !fIsTransferred) {
+                   if(ptr) delete [] ptr; // Previously it deletes ptr without verifying ptr. It causes double free memory. Need more examination.
+
+                   fUnzipChunks[reqi] = 0;
+                   fUnzipLen[reqi] = 0;
+                   fUnzipStatus[reqi] = 2; // Set it as not done
 
-            if (fUnzipStatus) delete [] fUnzipStatus;
-            if (fUnzipLen) delete [] fUnzipLen;
-            if (fUnzipChunks) delete [] fUnzipChunks;
+                   return 1;
+           }
 
-            fUnzipStatus  = aUnzipStatus;
-            fUnzipLen  = aUnzipLen;
-            fUnzipChunks = aUnzipChunks;
+           fUnzipChunks[reqi] = ptr;
+           fUnzipLen[reqi] = loclen;
+           fUnzipStatus[reqi] = 2; // fUnzipStatus[reqi] is updated after fUnzipChunks[reqi] and fUnzipLen[reqi] get updated because we need to synchronize the main thread in GetUnzipBuffer().
 
-            fNseekMax  = fNseek;
+           fNUnzip++;
+   }
+   else {
+           fUnzipChunks[reqi] = 0;
+           fUnzipLen[reqi] = 0;
+           fUnzipStatus[reqi] = 2;
+   }
+
+   return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// We create a TTaskGroup and asynchronously maps each group of baskets(> 100 kB in total)
+/// to a task. In TTaskGroup, we use TThreadExecutor to do the actually work of unzipping 
+/// a group of basket. The purpose of creating TTaskGroup is to avoid competing with main thread.
+
+Int_t TTreeCacheUnzip::CreateTasks()
+{
+   auto mapFunction = [&]() {
+      auto unzipFunction = [&](const std::vector<Int_t> &indices) {
+         for (auto ii : indices) {
+            Byte_t oldV = 0;
+            Byte_t newV = 1;
+            if(fUnzipStatus[ii].compare_exchange_weak(oldV, newV, std::memory_order_release, std::memory_order_relaxed)) {
+               Int_t locbuffsz = 16384;
+               char *locbuff = new char[16384];
+               Int_t res = UnzipCache(ii, locbuffsz, locbuff);
+               if(res)
+                  if (gDebug > 0)
+                     Info("UnzipCache", "Unzipping failed or cache is in learning state");
+            }
          }
+         return nullptr;
+      };
+
+      Int_t accusz = 0;
+      std::vector<std::vector<Int_t>> basketIndices;
+      std::vector<Int_t> indices;
+      for (Int_t i = 0; i < fNseek; i++) {
+         while (accusz < 102400) {
+            accusz += fSeekLen[i];
+            indices.push_back(i);
+            i++;
+            if (i >= fNseek) break;
+         }
+         if (i < fNseek) i--;
+         basketIndices.push_back(indices);
+         indices.clear();
+         accusz = 0;
+      }
+      ROOT::TThreadExecutor pool;
+      pool.Foreach(unzipFunction, basketIndices);
```

Comment:
I think this is the source of your speed issues.

Using `TThreadExecutor` causes you to have to pre-create all the tasks; the first one isn't executed until all are created.  Use the `TTaskGroup` object and it will schedule the tasks as they are created.

Reply:
As we discussed earlier, the performance was decremented by another ~3% if I use TTaskGroup.Run here.

Since I still keep outer TTaskGroup, so there is still ~3% performance drop, but replacing inner TTaskGroup by TThreadExecutor mitigate the slowdown.

Reply:
Ok - per our discussion at the ROOT IO meeting, let's leave this for now and just focus on the code style cleanups.